sysl2/sysl/seqs.go

```diff
@@ -0,0 +1,1081 @@
+}
+
+type sequenceDiagWriter struct {
+	autogenHeader bool
+	head          indentWriter
+	body          indentWriter
+}
+
+func (w *sequenceDiagWriter) String() string {
+	var sb strings.Builder
+	if w.autogenHeader {
+		fmt.Fprintln(&sb, "''''''''''''''''''''''''''''''''''''''''''")
+		fmt.Fprintln(&sb, "''                                      ''")
+		fmt.Fprintln(&sb, "''  AUTOGENERATED CODE -- DO NOT EDIT!  ''")
+		fmt.Fprintln(&sb, "''                                      ''")
+		fmt.Fprintln(&sb, "''''''''''''''''''''''''''''''''''''''''''")
+		fmt.Fprintln(&sb)
+	}
+
+	fmt.Fprintln(&sb, "@startuml")
+	sb.WriteString(w.head.Buffer.String())
+	sb.WriteString(w.body.Buffer.String())
+	fmt.Fprintln(&sb, "@enduml")
+
+	return sb.String()
+}
+
+type SimpleParser struct {
+	self string
+}
+
+type epFmtParam struct {
+	epname, human, human_sender, needs_int, args, patterns, controls string
+	attrs                                                            map[string]*sysl.Attribute
+}
+
+type SfmtApp = func(appname, controls string, attrs map[string]*sysl.Attribute) string
+
+type SfmtEP = func(p *epFmtParam) string
+
+type sequenceDiagParam struct {
+	endpoints   []string
+	epfmt       SfmtEP
+	appfmt      SfmtApp
+	activations bool
+	title       string
+	blackboxes  [][]string
+}
+
+type appEndpoint struct {
+	appname, epname, upto string
+}
+
+func makeAppEndpoint(s string) *appEndpoint {
+	r := regexp.MustCompile(`(?P<appname>.*?)\s*<-\s*(?P<epname>.*?)(?:\s*\[upto\s+(?P<upto>.*)\])*$`)
+	match := r.FindStringSubmatch(s)
+
+	out := &appEndpoint{}
+	for i, name := range r.SubexpNames() {
+		if i > 0 && i <= len(match) {
+			switch {
+			case name == "appname":
+				out.appname = match[i]
+			case name == "epname":
+				out.epname = match[i]
+			case name == "upto":
+				out.upto = match[i]
+			}
+		}
+	}
+	return out
+}
+
+type varName struct {
+	category, order       int
+	agent, varname, label string
+}
+
+func (v *varName) String() string {
+	return fmt.Sprintf(`%s "%s" as %s`, v.agent, v.label, v.varname)
+}
+
+func (v *varName) setCategoryAndAgent(attrs map[string]*sysl.Attribute) {
+	if attrs != nil {
+		if patterns, has := attrs["patterns"]; has {
+			if x := patterns.GetA(); x != nil {
+				for _, y := range x.Elt {
+					switch y.GetS() {
+					case "human":
+						v.category = 0
+						v.agent = "actor"
+					case "ui":
+						v.category = 1
+						v.agent = "boundary"
+					case "cron":
+						v.category = 2
+						v.agent = "control"
+					case "db":
+						v.category = 4
+						v.agent = "database"
+					case "external":
+						v.category = 5
+						v.agent = "control"
+					}
+				}
+			}
+		}
+	}
+}
+
+func makeVarName(varname string) *varName {
+	i, _ := strconv.Atoi(varname[1:])
+	return &varName{3, i, "control", varname, ""}
+}
+
+type newVar = func(varname, appname string)
+
+type varManager struct {
+	symbols       map[string]string
+	newVarCreated newVar
+}
+
+func (v *varManager) getOrCreateVar(appname string) string {
```

Comment:
Call it `uniqueVarForAppName`. Whether it's created in the process or not is immaterial. Also, generally avoid the prefix `get`.